nexus/src/nexus.rs

```diff
@@ -1056,35 +1048,154 @@ impl Nexus {
+                    .ok_or("no content length!")
+                    .map_err(|e| Error::InvalidValue {
+                        label: String::from("url"),
+                        message: format!("error querying url: {}", e),
+                    })?;
+
+                let total_size =
+                    u64::from_str(content_length.to_str().map_err(|e| {
+                        Error::InvalidValue {
+                            label: String::from("url"),
+                            message: format!("content length invalid: {}", e),
+                        }
+                    })?)
+                    .map_err(|e| {
+                        Error::InvalidValue {
+                            label: String::from("url"),
+                            message: format!("content length invalid: {}", e),
+                        }
+                    })?;
+
+                let size: external::ByteCount = total_size.try_into().map_err(
+                    |e: external::ByteCountRangeError| Error::InvalidValue {
+                        label: String::from("size"),
+                        message: format!("total size is invalid: {}", e),
+                    },
+                )?;
+
+                // validate total size is divisible by block size
+                let block_size: u64 = params.block_size.into();
+                if (size.to_bytes() % block_size) != 0 {
+                    return Err(Error::InvalidValue {
+                        label: String::from("size"),
+                        message: format!(
+                            "total size {} must be divisible by block size {}",
+                            size.to_bytes(),
+                            block_size
+                        ),
+                    });
+                }
+
+                // for images backed by a url, store the ETag as the version
+                let etag = response
+                    .headers()
+                    .get(reqwest::header::ETAG)
+                    .map(|x| x.to_str().ok())
+                    .flatten()
+                    .map(|x| x.to_string());
+
+                let new_image_volume =
+                    db::model::Volume::new(Uuid::new_v4(), volume_data);
+                let volume =
+                    self.db_datastore.volume_create(new_image_volume).await?;
+
+                db::model::GlobalImage {
+                    identity: db::model::GlobalImageIdentity::new(
+                        Uuid::new_v4(),
+                        params.identity.clone(),
+                    ),
+                    volume_id: volume.id(),
+                    url: Some(url.clone()),
+                    version: etag,
+                    digest: None, // not computed for URL type
+                    block_size: db_block_size,
+                    size: size.into(),
+                }
+            }
+
+            params::ImageSource::Snapshot(_id) => {
+                unimplemented!()
```

Comment:
How about returning a 500 here instead of panicking?  (You could even use `unimplemented_todo`, but it's not necessary.)

Reply:
opted for 503 in 3b883917

Reply:
I think using 503 for this case makes it harder for us to identify operational issues (usually 503s) vs. bugs (usually 500s) in the field...but hopefully this will never ship so it won't matter!